tools/tck-integrity-tests/src/test/scala/org/opencypher/tools/tck/ValidateSteps.scala

```diff
@@ -0,0 +1,138 @@
+        numberOfExpectResultSteps,
+        numberOfExplicitSideEffectSteps,
+        numberOfImplicitSideEffectSteps = 0
+    var positionLastInitQuery = Int.MinValue
+    var positionFirstExecQuery = Int.MaxValue
+    steps.zipWithIndex foreach {
+      case (Execute(_, InitQuery, _), ix) => positionLastInitQuery = Math.max(positionLastInitQuery, ix)
+      case (Execute(_, ExecQuery, _), ix) => numberOfExecQuerySteps += 1
+        positionFirstExecQuery = Math.min(positionFirstExecQuery, ix)
+      case (Execute(_, ControlQuery, _), _) => numberOfControlQuerySteps += 1
+      case (_: ExpectError, _) => numberOfExpectErrorSteps += 1
+      case (_: ExpectResult, _) => numberOfExpectResultSteps += 1
+      case (se: SideEffects, _) if se.source.getType == StepType.AND => numberOfExplicitSideEffectSteps += 1
+      case (se: SideEffects, _) if se.source.getType == StepType.THEN => numberOfImplicitSideEffectSteps += 1
+      case _ => Unit
+    }
+    withClue("scenario has exactly one `When executing query` steps") {
+      numberOfExecQuerySteps shouldBe 1
+    }
+
+    withClue("scenario has at most one `When executing control query` steps") {
+      numberOfControlQuerySteps should be <= 1
+    }
+
+    withClue("scenario has at most one `Then expect error` steps") {
+      numberOfExpectErrorSteps should be <= 1
+    }
+
+    withClue("scenario has as many `Then expect results` and `Then expect error` steps as `When executing query` and `When executing control query` steps") {
+      (numberOfExpectResultSteps + numberOfExpectErrorSteps) should equal(numberOfExecQuerySteps + numberOfControlQuerySteps)
+    }
+
+    withClue("scenario has as many explicit `And side effects` step or implicit SideEffects step as `When executing query` steps") {
+      (numberOfExplicitSideEffectSteps + numberOfImplicitSideEffectSteps) shouldBe numberOfExecQuerySteps
+    }
+
+    withClue("scenario start with a `Given` step") {
+      steps.head.source.getType shouldBe StepType.GIVEN
+    }
+
+    (steps zip steps.tail) foreach {
+      case (predecessor, er: ExpectResult) =>
+        withClue(s"${er.description} is preceded by a `When executing query` or `When executing control query` step") {
+          predecessor should matchPattern { case Execute(_, ExecQuery | ControlQuery, _) => }
+        }
+      case (predecessor, ee: ExpectError) =>
+        withClue(s"${ee.description} is preceded by a `When executing query` step") {
+          predecessor should matchPattern { case Execute(_, ExecQuery, _) => }
+        }
+      case (predecessor, se: SideEffects) if se.source.getType == StepType.AND =>
+        withClue(s"${se.description} is preceded by a `Then expect results` step") {
+          predecessor should matchPattern { case _: ExpectResult => }
+        }
+      case (predecessor, se: SideEffects) if se.source.getType == StepType.THEN =>
+        withClue(s"${se.description} is preceded by a `Then expect error` step") {
```

Comment:
I think there should be constants for these step phrases (`Then expect error` et al)

Reply:
Hm, only parts of it: https://github.com/opencypher/openCypher/blob/master/tools/tck-api/src/main/scala/org/opencypher/tools/tck/constants/TCKStepDefinitions.scala